### DISCUSSION TITLE
Meeting notes: Add new notes for Feb 26, March 9

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#3302] [General]
March 9, 2020

[C# Language Design Notes for March 9, 2020](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-03-09.md)

 1.   Simple programs

  2.  Records

Feb 26, 2020

[C# Language Design Notes for Feb. 26, 2020](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-02-26.md)

Design Review

**Comment by YairHalberstadt** [comment]
I believe the links are the wrong way round

**Comment by Joe4evr** [comment]
> * One of the biggest drawbacks of the current writeable-property style in C#, where types are declared with public mutable properties that are then initialized using object initializers, is that author can't enforce that certain properties are always initialized. It would be a big disappointment if any "nominal records" feature that we built couldn't support this feature.

Once again, I'd like to point the LDM's attention to #2328 which can simply be another building block.

**Comment by agocke** [comment]
> I believe the links are the wrong way round

Fixed!

> #2328 

Good to know. Personally I'm not sure I buy a separate feature here. I think the concept is valid for init-only properties, but properties with public setters can have any value set at any time, so I'm not convinced it's useful.

**Comment by MgSam** [comment]
I still don't see what problem you guys are trying to solve with "simple programs". As you stated clearly in the design notes- it's impossible to reconcile with CSX. So is introducing a 3rd variant of C# really worth all the added complexity for the tiny marginal benefit? What is wrong with CSX exactly?

Are "learners" really an important target audience? C# is 20 years old and has consistently been one of the top used languages. Obviously, a lot of people have learned how to use it without requiring top-level statements.

**Comment by 333fred** [comment]
>  I still don't see what problem you guys are trying to solve with "simple programs".

Jupyter notebook scripting and other top-level programs are starting to become more popular. We want to make the transition between a notebook and a full C# program as seamless as possible, namely copy-paste with no additional work. We think that's well worth the costs. Personally, I'm also quite excited by the potential for a real front-and-center single file solution in `dotnet` itself, that's indistinguishable from `dotnet run`. Because it literally is `dotnet run`.

**Comment by alrz** [comment]
I was thinking this could simplify `Program.cs` for .net hosted applications (aspnet, workers, etc), but in the simplest case we need `args` argument. 

Would it be possible to define `void Main(string[] args)` as a top-level method?

**Comment by agocke** [comment]
You could probably use `Environment.CommandLine` for that case

**Comment by HaloFour** [comment]
@agocke 

That would require some portion of that "simple program" be dedicated to parsing that single string of args.  VB5/6 had similar limitations and it was a common/constant source of bugs, particularly given that (at the time) the string could be formatted differently on different OSes (9x vs. NT).

If C# isn't going to offer the args I would say that the BCL should add a method that would make it easy to obtain the args parsed correctly.
